### ISSUE TITLE
Analysis completion plugin callback

### ISSUE THREAD

**Issue by joshwatson** [Vector35/binaryninja-api#505]
It would be nice if there were a callback for analysis completion that plugins could register so that they could be run immediately after analysis completes, rather than requiring some sort of user interaction.

**Comment by Manouchehri** [comment]
Is there a way to launch a plugin automatically _before_ analysis starts?

**Comment by nbsdx** [comment]
@Manouchehri My UI plugins will "run" as soon as the page is created, but I don't know if they complete before analysis starts (because of threading, and I don't know what the internal order of events is for opening a new binary). But that's being done by injecting event filters from Qt, rather than through the official API, so I can only react to UI events rather than internal Binary Ninja events.

**Comment by plafosse** [comment]
Are you requesting something other than UpdateAnalysisAndWait() or was this just before that API existed?

**Comment by joshwatson** [comment]
It's a request for the ability to register a plugin to run automatically after analysis finishes in the UI, not headless.

I don't honestly remember for what purpose I initially requested this, but it was mostly so a user could piggyback off auto analysis as soon as a binary is opened, or even when auto analysis reruns at any point in the UI

**Comment by joshwatson** [comment]
I was chatting with Brian the other day, and I mentioned to him what I kinda envision for this particular issue to be implemented. An analysis completion plugin like this would look something like this:

```python
def analysis_callback(completion_event):
    # do stuff

BinaryViewType['PE'].register_analysis_completion_event(analysis_callback)
```
And in binaryview.py, `BinaryViewType.create` would look something like (but probably not exactly like):
```python
	def create(self, data):
		view = core.BNCreateBinaryViewOfType(self.handle, data.handle)
		if view is None:
			return None
		bv = BinaryView(file_metadata=data.file, handle=view)
		callbacks = list(core.BNGetRegisteredAnalysisCompletionEvents(self.handle))
		for c in callbacks:
			bv.add_analysis_completion_event(c)
		return bv
```

So this would allow a user to register an analysis completion event to run automatically whenever a BinaryView of that type were opened without having to manually run a plugin every time.

**Comment by gaasedelen** [comment]
There was some more discussion on Slack regarding this issue today, and I'd like to voice my support for this feature to come sooner than later.

Besides enabling a plugin to perform additional (automated) database augmentation *(e.g. Linear Sweep, PDB/Symbols)*, it's a necessary feature for building deeply integrated experiences *(c.f. Ethrays, Lighthouse, Collaboration ...)* and exposing them naturally to the user.

For robustness, it would be nice to have a plugin accessible event/callback when that trigggers when a BNDB/binary is opened by the UI (6. in #1097), and after the initial BNDB analysis has completed (when BN becomes idle).

I am not sure how you might schedule compeiting user-plugins wishing to perform their own analyses after the initial core BN analysis has finished. This is probably a less common case though.

**Comment by sraboy** [comment]
I want to +1 this but also +1 @Manouchehri's question. I'd like a callback when analysis is completed but I'd also like the ability for a plugin to be running its thread as soon as Binja launches, so I can pass in a filepath or `BinaryView` to Binja.

My particular use case is to provide a UI for a user to pull/open files from a Docker container or S3 bucket from within the UI.

**Comment by jeffli678** [comment]
Implemented in dev 2.2.2528. A minimal demo is:

```Python
def bv_finalized_callback(bv):
   # or do anything with the bv
   bv.store_metadata('finalized', 'yes')

BinaryViewType.add_binaryview_finalized_event(bv_finalized_callback)
bv = BinaryViewType.get_view_of_file('/bin/ls')


```
